README.md
# Tarot, but Game

<p style="text-align:center;">
    The CSS Crusaders (Team 17)<br>
    <img src="./admin/branding/mascot.jpeg" alt="CSS Crusaders Mascot" style="width:50px;"/>
</p>

[Tarot, but Game (link to web version of the project)](https://cse110-team17.github.io/cse110-sp23-group17/source/)

[JSDocumentation](https://cse110-team17.github.io/cse110-sp23-group17/out/)

[Our Team Page](/admin/team.md) | [Our Team Video](https://www.youtube.com/watch?v=jR47GaIusic&ab_channel=KristinE)

[Private video](https://youtu.be/FqfbOe_-QS0)

[Public video](https://youtu.be/BD69aKURB_E)

## Project Overview
![last commit in main](https://img.shields.io/github/last-commit/CSE110-Team17/cse110-sp23-group17/main)
![open issues](https://img.shields.io/github/issues-raw/CSE110-Team17/cse110-sp23-group17)
![open pull requests](https://img.shields.io/github/issues-pr-raw/CSE110-Team17/cse110-sp23-group17)
![license](https://img.shields.io/github/license/CSE110-Team17/cse110-sp23-group17)

Welcome to Tarot, but Game repository! **Tarot, but Game** introduces a **gameplay experience** to getting **tarot card fortunes!** 

Our team created a Tarot card fortune telling web application that improves the user experience by making it **more fun and interesting**. You will be able to go on a journey where you can learn about tarot cards, choose your character and meet a tarot fortune teller who will reveal your fortune. Here, you will be able to find our source code in `./source/` and documentations throughout our repository. The overview of documentation locations are listed below.

### Documentation Location
- ADRs: `./specs/adrs`
- Team Meeting Notes: `./admin/meetings`
- Pitch Documents: `./specs/pitch`
- Brainstorming Documents `./specs/brainstrom`
- CI/CD Pipeline: `./.github/workflows` and `./admin/cipipeline`
- Videos: `./admin/vidoes`

## How to Locally Build and Test
### Clone the repo
```shell
# Clone repo
git clone git@github.com:CSE110-Team17/cse110-sp23-group17.git
# Change working directory
cd se110-sp23-group17
```

### Install all dependencies
```shell
# If you don't have node, check https://nodejs.dev/en/learn/how-to-install-nodejs/ to install
# How to check if you have node installed (MacOS)
node -v
# If you have node, install dependencies. This action will create package.json and package-lock.json
npm install
```

### Running Tests, JSDocs, ESLint
```shell
# Run Tests
npm run test

# Run JSDocs
npm run jsdoc

# Run ESLint
npm run lint 
```

### Building the web application
Use Visual Studio Code Live Server extension. If you don't have it you can install it [here](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer)

## How to Contribute
If you have a feature you wish to add or changes you wish to make, we are accepting suggestions on our [Feature Requests Form](https://tarot-but-game.canny.io/feature-requests)

## License

![license](https://img.shields.io/github/license/CSE110-Team17/cse110-sp23-group17)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Tarot but Game](https://github.com/CSE110-Team17/cse110-sp23-group17).

## [1.0] - 2023-6-10

### Added 
- added All Cards feature (displays all the possible cards you can get) (#69)
- added how-to-play button in game page so that the user can see game instructions anytime (#66)
- added card hovering animation in game play screen
- added card flip animation in game play screen
- added speaking fortune teller character
- added LICENSE
- added Feature Request Form
- added stars to home page
- added stars to camera page

### Fixed
- fixed cards overlapping with popup
- fixed buttons overlapping with luck meter

### Changed
- How to play instructions are changed to match the new gameplay (#66)
- changed the back button to match the design theme (#66)
- update README.md
- changed results page layout so all cards are in a row (#77)
- update ADRs
- changed stars in results page

### Removed
- removed How to Play button and Back button on mobile version of game play screen
- removed old game instructions
- removed language picker
- removed opponent and attack damage

## [0.4] - 2023-6-6

### Added
- added lucky points (Very Unclucky: 0-25, Unlucky: 25-50, Lucky: 50-75, Very Lucky: 75-100) (#68)
- added progress bar that go back and front (when click a card, the value of the progress bar will be the damage) (#54)
- added short-term fortune: lucky points ("How Your Luck Looks")
- added long-term fortune: card fortunes ("Your Overall Fortune")
- added back button to home page from game play screen

### Fixed
- fixed css hover effect on Camera Page (#61)
- fixed no image found when clicking next without any selector on Camera Page (#61)
- JSDoc error fix
- prevent clicking more cards than 4

### Changed
- Camera Page UI Design (#61)
- Camera Page now scalable to moble (#61)
- Update Unit Tests
- changed attack damage alert message to a modal box (#54)
- results page always displays 4 cards (#62)
- new card descriptions

### Removed
- removed "Card Position Represetation" (#62)

## [0.3] - 2023-5-30

### Added
- added CHANGELOG.md
- added Unit Tests Game Page (#29)
- added Unit Tests Results Page (#30)
- added JSDocs
- added oscillating bar that corresponds to attack damage
- can now select 2-4 cards depending on how much cards it takes to defeat the opponent

### Fixed
- Linting errors in JS files (#48)

### Changed
- increased the margins for results card descriptions (#51)

### Removed
- removed "Card Name" (#51)

## [0.2] - 2023-05-28

### Added
- deployed Tarot but Game to Github Pages
- Prettier formatting
- CI/CD Pipeline
- Eslint to CI/CD Pipeline
- added Sprint Retrospective
- Camera Page Outline (#31)
- JS for Camera Page (#31)
  - can take a picture with device camera or use an exisiting photo to change the user picture
- "Here is What Your Future Holds" text in Results Page (#28)
- added CONTRIBUTING.md

### Fixed

- fixed bug for images size (#23)
- fix bug when click the same card (#24)
- fixed button appearing in web version of Results Page
- fixed button display in mobile version of results page (#28)
- fixed readability in Results Page (#28) (improved margins)

### Changed

- make the Welcome Page fit mobile screen (#25)
- changed theme of Game Page (#27)
- make the Game Page fit mobile screen (#27)
- changed the layout of Results Page (#28) 
- changed theme of Results Page (#28)

## [0.1] - 2023-05-16

### Added

- Welcome Page Outline (#13)
  - title, subtitle, "how to play" and "play" button
- Instructions Page Outline (#4)
  - describe objective and rules of the game
- Game Page Outline (#1)
  - opponent/user stage areas, 22 cards, HP bar/display
- Results Page Outline (#2)
  - mobile version shows one card at a time with "next" and "previous" buttons to show a different card
- Tarot Card Information (#5)
- JS for Results Page (#7)
  - display cards and descriptions that is chosen in game play, allow "next" and "previous" buttons to work on mobile version
- JS for Game Page (#6)
  - generate random card when clicked, opponent HP decrease according to damage points, display game ending message when opponent HP is lower than 0
- Initial Theme Design in Welcome Page

source/assets/scripts/game.js
import tarotConfig from "../tarot.js";

window.addEventListener("DOMContentLoaded", init);

function init() {
  const board = document.querySelector(".board");
  const playerImage = document.querySelector(".player img");
  const luckLabel = document.querySelector(".luck-bar .label");
  const luckBarFill = document.querySelector(".luck-bar .fill");
  const oscillatingBar = document.querySelector(".oscillating-bar");
  const oscillatingBarFill = document.querySelector(".oscillating-bar .fill");
  const oracleMsg = document.querySelector(".oracle .message");

  playerImage.src = window.localStorage.getItem("userImage");

  let cardHTMLs = "";
  for (let i = 0; i < 22; i++) {
    cardHTMLs += 
    `<div class="card-container" id="card-container-${i + 1}">
      <div class="card" id="card-${i + 1}">
        <div class="back face"></div>
        <div class="front face"></div>
      </div>
    </div>`;
  }
  board.innerHTML = cardHTMLs;

  const cards = board.children;
  console.log(board.children);

  const startingLuck = 50;
  let luck;
  setLuck(startingLuck);

  //Function to load profile image
  window.loadFile = function (event) {
    var image = document.getElementById("output");
    image.src = URL.createObjectURL(event.target.files[0]);
  };

  //FOR RESULT PAGE: array of all the selected cards during game play
  const chosenCards = [];

  //listen whenever a card is clicked
  for (let i = 0; i < cards.length; i++) {
    cards[i].addEventListener("click", onCardClicked);
  }

  /**
   * Play the game when a card is click:
   * Generate random card name -> change image accoridngly
   * Generate random damage points -> change hp bar accordingly
   * @param {MouseEvent} event - the value return by the event
   *
   */
  function onCardClicked(event) {
    const card = event.currentTarget; // actually is card-container

    if (
      card.classList.contains("flipped") ||
      luck <= 0 ||
      luck >= 100 ||
      chosenCards.length >= 4
    ) {
      return;
    }

    card.classList.add("flipped");

    // 50-50 chance that the card is upside-down
    const isDown = Math.random() < 0.5;

    if (isDown) card.classList.add("reversed");

    if (isDown) {
      setLuck(
        Math.max(0, luck - getBarWidth(oscillatingBarFill, oscillatingBar))
      );
    } else {
      setLuck(
        Math.max(0, luck + getBarWidth(oscillatingBarFill, oscillatingBar))
      );
    }

    if (luck === 0 || luck === 100) {
      oscillatingBar.style.visibility = "hidden";
    }

    // Random generate a card. If card is already chosen then generate another card
    let randNameIdx = Math.floor(Math.random() * 21);
    let cardName = getTarotCardName(tarotConfig)[randNameIdx];
    while (chosenCards.includes(cardName)) {
      randNameIdx = Math.floor(Math.random() * 21);
      cardName = getTarotCardName(tarotConfig)[randNameIdx];
    }
    chosenCards.push(cardName);

    // Change the image according to the card got chosen
    tarotConfig.tarot.forEach((element) => {
      if (element.name === cardName) {
        card.querySelector(
          ".front"
        ).style.backgroundImage = `url("${element.image}")`;
      }
    });

    // Random generate a damage point and attack the oponent with that point.
    // Change the hp bar of opponent accordingly.
    if (isDown) {
      say(
        "You got a reverse " +
          cardName +
          " card. You receive " +
          getBarWidth(oscillatingBarFill, oscillatingBar) * -1 +
          " luck points!"
      );
    } else {
      say(
        "You got a " +
          cardName +
          " card. You receive " +
          getBarWidth(oscillatingBarFill, oscillatingBar) +
          " luck points!"
      );
    }

    setTimeout(() => {
      if (chosenCards.length === 4) {
        say("Get ready to see your fortune!");
        setTimeout(() => {
          localStorage.setItem("chosenCards", JSON.stringify(chosenCards));
          localStorage.setItem("luck", luck);
          window.location.href = "./results.html";
        }, 3000);
      }
    }, 3000);
  }

  let msgResetTimeout = -1;

  function say(msg) {
    oracleMsg.innerText = msg;

    clearTimeout(msgResetTimeout);

    msgResetTimeout = setTimeout(() => {
      const numCardsLeft = 4 - chosenCards.length;
      oracleMsg.innerText = `Draw ${numCardsLeft} more card${
        numCardsLeft === 1 ? "" : "s"
      }!`;
    }, 3000);
  }

  function setLuck(val) {
    if (val > 100) val = 100;
    if (val < 0) val = 0;
    luck = val;
    luckLabel.innerText = `${luck} luck points`;
    luckBarFill.style.width = `${luck}%`;
  }
}
export function getBarWidth(oscillatingBarFill, oscillatingBar) {
  const barWidth = oscillatingBarFill.getBoundingClientRect().width;
  const parentWidth = oscillatingBar.getBoundingClientRect().width;
  return Math.floor((15 * barWidth) / parentWidth);
}
/**
 * Create an array of 22 and parse all the card name from json file
 * @param {Array<*>} tarotConfig - an array of the tarot cards from json
 * @return {Array<string>} array contains 22 tarot cards' name
 */
export function getTarotCardName(tarotConfig) {
  const tarotCardNames = [];
  tarotConfig["tarot"].forEach((element) => {
    tarotCardNames.push(element["name"]);
  });
  return tarotCardNames;
}
